File: scripts/flux_tray_linux.py
#!/usr/bin/env python3
"""
FluxDefense Linux System Tray
A system tray application for Linux that displays real-time system metrics
"""

import gi
gi.require_version('Gtk', '3.0')
gi.require_version('AppIndicator3', '0.1')
from gi.repository import Gtk, AppIndicator3, GLib
import json
import subprocess
import threading
import time
import os
import sys

class SystemMetrics:
    def __init__(self):
        self.cpu_usage = 0.0
        self.memory_usage = 0.0
        self.memory_used_gb = 0.0
        self.network_rx_rate = 0.0
        self.network_tx_rate = 0.0
        self.disk_read_rate = 0.0
        self.disk_write_rate = 0.0
        self.load_average = [0.0, 0.0, 0.0]
        self.process_count = 0
        self.uptime_seconds = 0
        self.last_net_rx = 0
        self.last_net_tx = 0
        self.last_disk_read = 0
        self.last_disk_write = 0
        self.last_update = time.time()

    def update(self):
        """Update metrics from flux-monitor or /proc"""
        try:
            # Try flux-monitor first
            flux_path = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), 
                                   'target', 'release', 'flux-monitor')
            if os.path.exists(flux_path):
                result = subprocess.run([flux_path, 'metrics', '--json', '--once'], 
                                      capture_output=True, text=True, timeout=5)
                if result.returncode == 0:
                    data = json.loads(result.stdout)
                    self.cpu_usage = data.get('cpu_usage', 0.0)
                    self.memory_usage = data.get('memory_usage', 0.0)
                    self.memory_used_gb = data.get('memory_used_gb', 0.0)
                    self.network_rx_rate = data.get('network_rx_rate', 0.0)
                    self.network_tx_rate = data.get('network_tx_rate', 0.0)
                    self.disk_read_rate = data.get('disk_read_rate', 0.0)
                    self.disk_write_rate = data.get('disk_write_rate', 0.0)
                    self.load_average = data.get('load_average', [0.0, 0.0, 0.0])
                    self.process_count = data.get('process_count', 0)
                    self.uptime_seconds = data.get('uptime_seconds', 0)
                    return
        except Exception as e:
            print(f"Error getting metrics from flux-monitor: {e}")

        # Fallback to manual parsing
        self._update_from_proc()

    def _update_from_proc(self):
        """Update metrics by reading /proc files directly"""
        current_time = time.time()
        time_delta = current_time - self.last_update
        
        # CPU usage
        try:
            with open('/proc/stat', 'r') as f:
                line = f.readline()
                cpu_times = list(map(int, line.split()[1:8]))
                total = sum(cpu_times)
                idle = cpu_times[3] + cpu_times[4]  # idle + iowait
                self.cpu_usage = ((total - idle) / total) * 100 if total > 0 else 0
        except:
            pass

        # Memory usage
        try:
            with open('/proc/meminfo', 'r') as f:
                meminfo = {}
                for line in f:
                    parts = line.split()
                    if len(parts) >= 2:
                        meminfo[parts[0].rstrip(':')] = int(parts[1])
                
                total = meminfo.get('MemTotal', 1)
                available = meminfo.get('MemAvailable', 0)
                used = total - available
                self.memory_usage = (used / total) * 100 if total > 0 else 0
                self.memory_used_gb = used / 1024 / 1024
        except:
            pass

        # Network stats
        try:
            with open('/proc/net/dev', 'r') as f:
                lines = f.readlines()[2:]  # Skip headers
                total_rx = 0
                total_tx = 0
                for line in lines:
                    if ':' in line:
                        iface, stats = line.split(':', 1)
                        iface = iface.strip()
                        # Skip loopback and virtual interfaces
                        if iface not in ['lo', 'docker0', 'virbr0']:
                            values = stats.split()
                            if len(values) >= 9:
                                total_rx += int(values[0])
                                total_tx += int(values[8])
                
                if time_delta > 0:
                    self.network_rx_rate = (total_rx - self.last_net_rx) / time_delta if self.last_net_rx > 0 else 0
                    self.network_tx_rate = (total_tx - self.last_net_tx) / time_delta if self.last_net_tx > 0 else 0
                self.last_net_rx = total_rx
                self.last_net_tx = total_tx
        except:
            pass

        # Disk stats
        try:
            with open('/proc/diskstats', 'r') as f:
                total_read = 0
                total_write = 0
                for line in f:
                    parts = line.split()
                    if len(parts) >= 10:
                        device = parts[2]
                        # Only count physical disks
                        if device.startswith('sd') or device.startswith('nvme'):
                            if not any(c.isdigit() for c in device[-1:]):  # Skip partitions
                                total_read += int(parts[5]) * 512  # sectors to bytes
                                total_write += int(parts[9]) * 512
                
                if time_delta > 0:
                    self.disk_read_rate = (total_read - self.last_disk_read) / time_delta if self.last_disk_read > 0 else 0
                    self.disk_write_rate = (total_write - self.last_disk_write) / time_delta if self.last_disk_write > 0 else 0
                self.last_disk_read = total_read
                self.last_disk_write = total_write
        except:
            pass

        # Load average
        try:
            with open('/proc/loadavg', 'r') as f:
                parts = f.read().split()
                self.load_average = [float(parts[i]) for i in range(3)]
        except:
            pass

        # Process count
        try:
            proc_count = sum(1 for name in os.listdir('/proc') if name.isdigit())
            self.process_count = proc_count
        except:
            pass

        # Uptime
        try:
            with open('/proc/uptime', 'r') as f:
                self.uptime_seconds = int(float(f.read().split()[0]))
        except:
            pass

        self.last_update = current_time


class FluxDefenseTray:
    def __init__(self):
        self.metrics = SystemMetrics()
        self.indicator = AppIndicator3.Indicator.new(
            "fluxdefense-tray",
            "security-high",
            AppIndicator3.IndicatorCategory.SYSTEM_SERVICES
        )
        self.indicator.set_status(AppIndicator3.IndicatorStatus.ACTIVE)
        
        # Create menu
        self.menu = Gtk.Menu()
        
        # Status item
        self.status_item = Gtk.MenuItem(label="FluxDefense System Monitor")
        self.status_item.set_sensitive(False)
        self.menu.append(self.status_item)
        
        self.menu.append(Gtk.SeparatorMenuItem())
        
        # Dashboard item
        dashboard_item = Gtk.MenuItem(label="Open Dashboard")
        dashboard_item.connect("activate", self.open_dashboard)
        self.menu.append(dashboard_item)
        
        # Refresh item
        refresh_item = Gtk.MenuItem(label="Refresh Now")
        refresh_item.connect("activate", self.refresh_metrics)
        self.menu.append(refresh_item)
        
        self.menu.append(Gtk.SeparatorMenuItem())
        
        # About item
        about_item = Gtk.MenuItem(label="About")
        about_item.connect("activate", self.show_about)
        self.menu.append(about_item)
        
        # Quit item
        quit_item = Gtk.MenuItem(label="Quit")
        quit_item.connect("activate", self.quit)
        self.menu.append(quit_item)
        
        self.menu.show_all()
        self.indicator.set_menu(self.menu)
        
        # Start update thread
        self.running = True
        self.update_thread = threading.Thread(target=self.update_loop, daemon=True)
        self.update_thread.start()
        
        # Schedule UI updates
        GLib.timeout_add_seconds(3, self.update_ui)

    def get_status_color(self, value, warning, critical):
        """Get color emoji based on threshold"""
        if value >= critical:
            return "🔴"
        elif value >= warning:
            return "🟡"
        else:
            return "🟢"

    def format_rate(self, bytes_per_sec):
        """Format network/disk rate"""
        if bytes_per_sec < 1024:
            return "●"
        elif bytes_per_sec < 1024 * 1024:
            return f"{int(bytes_per_sec / 1024)}K"
        else:
            return f"{int(bytes_per_sec / 1024 / 1024)}M"

    def update_ui(self):
        """Update the UI with current metrics"""
        m = self.metrics
        
        # Create status text
        cpu_color = self.get_status_color(m.cpu_usage, 70, 90)
        mem_color = self.get_status_color(m.memory_usage, 80, 95)
        
        net_status = f"{self.format_rate(max(m.network_rx_rate, m.network_tx_rate))}🟢"
        disk_status = f"{self.format_rate(max(m.disk_read_rate, m.disk_write_rate))}🟢"
        
        status_text = f"CPU{cpu_color}{int(m.cpu_usage)}% RAM{mem_color}{int(m.memory_usage)}% ({m.memory_used_gb:.1f}GB) NET{net_status} DISK{disk_status}"
        
        # Update label
        self.indicator.set_label(status_text, "FluxDefense")
        
        # Create tooltip
        hours = m.uptime_seconds // 3600
        tooltip = f"""FluxDefense System Monitor
────────────────────────
CPU Usage: {m.cpu_usage:.1f}%
Memory: {m.memory_usage:.1f}% ({m.memory_used_gb:.2f} GB used)
Network: ↓{m.network_rx_rate/1024/1024:.1f} MB/s ↑{m.network_tx_rate/1024/1024:.1f} MB/s
Disk: R{m.disk_read_rate/1024/1024:.1f} MB/s W{m.disk_write_rate/1024/1024:.1f} MB/s
Load: {m.load_average[0]:.2f} {m.load_average[1]:.2f} {m.load_average[2]:.2f}
Processes: {m.process_count}
Uptime: {hours} hours"""
        
        self.status_item.set_label(tooltip)
        
        return True  # Continue timeout

    def update_loop(self):
        """Background thread to update metrics"""
        while self.running:
            self.metrics.update()
            time.sleep(3)

    def open_dashboard(self, widget):
        """Open the dashboard in a browser"""
        try:
            subprocess.run(["xdg-open", "http://localhost:8080"], check=False)
        except Exception as e:
            print(f"Failed to open dashboard: {e}")

    def refresh_metrics(self, widget):
        """Force refresh metrics"""
        self.metrics.update()
        self.update_ui()

    def show_about(self, widget):
        """Show about dialog"""
        dialog = Gtk.MessageDialog(
            parent=None,
            flags=0,
            message_type=Gtk.MessageType.INFO,
            buttons=Gtk.ButtonsType.OK,
            text="FluxDefense System Tray"
        )
        dialog.format_secondary_text(
            "Real-time system monitoring for Linux\n\n"
            "Part of the FluxDefense security suite\n"
            "https://github.com/fluxdefense/fluxdefense"
        )
        dialog.run()
        dialog.destroy()

    def quit(self, widget):
        """Quit the application"""
        self.running = False
        Gtk.main_quit()


def main():
    # Check if we're on Linux
    if sys.platform != 'linux':
        print("This system tray implementation is Linux-specific.")
        print("For macOS, use the FluxDefenseUI Swift application.")
        sys.exit(1)
    
    # Check for required dependencies
    try:
        import gi
        gi.require_version('Gtk', '3.0')
        gi.require_version('AppIndicator3', '0.1')
    except Exception as e:
        print("Error: Missing required dependencies")
        print("Please install: sudo apt-get install python3-gi gir1.2-appindicator3-0.1")
        sys.exit(1)
    
    # Create and run the tray application
    app = FluxDefenseTray()
    Gtk.main()


if __name__ == "__main__":
    main()

File: src/bin/flux_tray_linux.rs
use anyhow::Result;
use std::process::Command;
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Duration;
use serde::{Deserialize, Serialize};

#[cfg(target_os = "linux")]
use gtk::prelude::*;
#[cfg(target_os = "linux")]
use libappindicator::{AppIndicator, AppIndicatorStatus};

#[derive(Debug, Clone, Serialize, Deserialize)]
struct SystemMetrics {
    cpu_usage: f64,
    memory_usage: f64,
    memory_used_gb: f64,
    network_rx_rate: f64,
    network_tx_rate: f64,
    disk_read_rate: f64,
    disk_write_rate: f64,
    load_average: Vec<f64>,
    process_count: u32,
    uptime_seconds: u64,
}

impl Default for SystemMetrics {
    fn default() -> Self {
        Self {
            cpu_usage: 0.0,
            memory_usage: 0.0,
            memory_used_gb: 0.0,
            network_rx_rate: 0.0,
            network_tx_rate: 0.0,
            disk_read_rate: 0.0,
            disk_write_rate: 0.0,
            load_average: vec![0.0; 3],
            process_count: 0,
            uptime_seconds: 0,
        }
    }
}

fn get_system_metrics() -> Result<SystemMetrics> {
    // Try to get metrics from flux-monitor first
    let output = Command::new("./target/release/flux-monitor")
        .args(&["metrics", "--json", "--once"])
        .output();

    if let Ok(output) = output {
        if output.status.success() {
            let json_str = String::from_utf8_lossy(&output.stdout);
            if let Ok(metrics) = serde_json::from_str::<SystemMetrics>(&json_str) {
                return Ok(metrics);
            }
        }
    }

    // Fallback to reading from /proc
    let mut metrics = SystemMetrics::default();

    // CPU usage from /proc/stat
    if let Ok(stat) = std::fs::read_to_string("/proc/stat") {
        let first_line = stat.lines().next().unwrap_or("");
        let values: Vec<u64> = first_line
            .split_whitespace()
            .skip(1)
            .filter_map(|v| v.parse().ok())
            .collect();
        
        if values.len() >= 4 {
            let total = values.iter().sum::<u64>();
            let idle = values[3];
            if total > 0 {
                metrics.cpu_usage = ((total - idle) as f64 / total as f64) * 100.0;
            }
        }
    }

    // Memory from /proc/meminfo
    if let Ok(meminfo) = std::fs::read_to_string("/proc/meminfo") {
        let mut total_kb = 0u64;
        let mut available_kb = 0u64;
        
        for line in meminfo.lines() {
            if line.starts_with("MemTotal:") {
                total_kb = line.split_whitespace().nth(1).and_then(|v| v.parse().ok()).unwrap_or(0);
            } else if line.starts_with("MemAvailable:") {
                available_kb = line.split_whitespace().nth(1).and_then(|v| v.parse().ok()).unwrap_or(0);
            }
        }
        
        if total_kb > 0 {
            let used_kb = total_kb - available_kb;
            metrics.memory_usage = (used_kb as f64 / total_kb as f64) * 100.0;
            metrics.memory_used_gb = used_kb as f64 / 1024.0 / 1024.0;
        }
    }

    // Load average from /proc/loadavg
    if let Ok(loadavg) = std::fs::read_to_string("/proc/loadavg") {
        let parts: Vec<f64> = loadavg
            .split_whitespace()
            .take(3)
            .filter_map(|v| v.parse().ok())
            .collect();
        metrics.load_average = parts;
    }

    // Process count
    if let Ok(entries) = std::fs::read_dir("/proc") {
        metrics.process_count = entries
            .filter_map(|e| e.ok())
            .filter(|e| e.file_name().to_str().map(|s| s.chars().all(|c| c.is_numeric())).unwrap_or(false))
            .count() as u32;
    }

    Ok(metrics)
}

fn get_status_color(value: f64, warning_threshold: f64, critical_threshold: f64) -> &'static str {
    if value >= critical_threshold {
        "🔴"
    } else if value >= warning_threshold {
        "🟡"
    } else {
        "🟢"
    }
}

fn format_rate(bytes_per_sec: f64) -> String {
    if bytes_per_sec < 1024.0 {
        "●".to_string()
    } else if bytes_per_sec < 1024.0 * 1024.0 {
        format!("{}K", (bytes_per_sec / 1024.0) as u32)
    } else {
        format!("{}M", (bytes_per_sec / 1024.0 / 1024.0) as u32)
    }
}

fn create_status_text(metrics: &SystemMetrics) -> String {
    let cpu_color = get_status_color(metrics.cpu_usage, 70.0, 90.0);
    let mem_color = get_status_color(metrics.memory_usage, 80.0, 95.0);
    
    let net_status = if metrics.network_rx_rate > 1024.0 || metrics.network_tx_rate > 1024.0 {
        format!("{}🟢", format_rate(metrics.network_rx_rate.max(metrics.network_tx_rate)))
    } else {
        "●🟢".to_string()
    };
    
    let disk_status = if metrics.disk_read_rate > 1024.0 || metrics.disk_write_rate > 1024.0 {
        format!("{}🟢", format_rate(metrics.disk_read_rate.max(metrics.disk_write_rate)))
    } else {
        "●🟢".to_string()
    };

    format!(
        "CPU{}{:.0}% RAM{}{:.0}% ({:.1}GB) NET{} DISK{}",
        cpu_color, metrics.cpu_usage,
        mem_color, metrics.memory_usage, metrics.memory_used_gb,
        net_status,
        disk_status
    )
}

fn create_tooltip_text(metrics: &SystemMetrics) -> String {
    format!(
        "FluxDefense System Monitor\n\
         ────────────────────────\n\
         CPU Usage: {:.1}%\n\
         Memory: {:.1}% ({:.2} GB used)\n\
         Network: ↓{:.1} MB/s ↑{:.1} MB/s\n\
         Disk: R{:.1} MB/s W{:.1} MB/s\n\
         Load: {:.2} {:.2} {:.2}\n\
         Processes: {}\n\
         Uptime: {} hours",
        metrics.cpu_usage,
        metrics.memory_usage, metrics.memory_used_gb,
        metrics.network_rx_rate / 1024.0 / 1024.0,
        metrics.network_tx_rate / 1024.0 / 1024.0,
        metrics.disk_read_rate / 1024.0 / 1024.0,
        metrics.disk_write_rate / 1024.0 / 1024.0,
        metrics.load_average.get(0).unwrap_or(&0.0),
        metrics.load_average.get(1).unwrap_or(&0.0),
        metrics.load_average.get(2).unwrap_or(&0.0),
        metrics.process_count,
        metrics.uptime_seconds / 3600
    )
}

#[cfg(target_os = "linux")]
fn main() -> Result<()> {
    // Initialize GTK
    gtk::init()?;

    // Create the app indicator
    let mut indicator = AppIndicator::new("fluxdefense-tray", "");
    indicator.set_status(AppIndicatorStatus::Active);
    indicator.set_icon("security-high");

    // Create menu
    let menu = gtk::Menu::new();
    
    // Status item (will be updated with metrics)
    let status_item = gtk::MenuItem::with_label("FluxDefense System Monitor");
    status_item.set_sensitive(false);
    menu.append(&status_item);
    
    menu.append(&gtk::SeparatorMenuItem::new());
    
    // Show Dashboard item
    let dashboard_item = gtk::MenuItem::with_label("Show Dashboard");
    let dashboard_item_clone = dashboard_item.clone();
    dashboard_item.connect_activate(move |_| {
        // Launch the dashboard in a browser
        if let Err(e) = Command::new("xdg-open")
            .arg("http://localhost:8080")
            .spawn()
        {
            eprintln!("Failed to open dashboard: {}", e);
        }
    });
    menu.append(&dashboard_item);
    
    menu.append(&gtk::SeparatorMenuItem::new());
    
    // Quit item
    let quit_item = gtk::MenuItem::with_label("Quit");
    quit_item.connect_activate(|_| {
        gtk::main_quit();
    });
    menu.append(&quit_item);
    
    menu.show_all();
    indicator.set_menu(&mut menu);

    // Shared metrics state
    let metrics = Arc::new(Mutex::new(SystemMetrics::default()));
    let metrics_clone = metrics.clone();

    // Start metrics update thread
    thread::spawn(move || {
        loop {
            if let Ok(new_metrics) = get_system_metrics() {
                if let Ok(mut m) = metrics_clone.lock() {
                    *m = new_metrics;
                }
            }
            thread::sleep(Duration::from_secs(3));
        }
    });

    // Update the display periodically
    let status_item_clone = status_item.clone();
    gtk::timeout_add(3000, move || {
        if let Ok(m) = metrics.lock() {
            let status_text = create_status_text(&m);
            let tooltip_text = create_tooltip_text(&m);
            
            status_item_clone.set_label(&status_text);
            
            // Set tooltip on the menu item
            if let Some(widget) = status_item_clone.get_child() {
                widget.set_tooltip_text(Some(&tooltip_text));
            }
        }
        gtk::Continue(true)
    });

    // Run GTK main loop
    gtk::main();
    
    Ok(())
}

#[cfg(not(target_os = "linux"))]
fn main() {
    eprintln!("This system tray implementation is Linux-specific.");
    eprintln!("For macOS, use the FluxDefenseUI Swift application.");
    std::process::exit(1);
}